src/main/java/uk/ac/ebi/ampt2d/accession/common/accessioning/BasicAccessioningService.java

```diff
@@ -0,0 +1,136 @@
+        this.accessionGenerator = accessionGenerator;
+        this.dbService = dbService;
+        this.digestFunction = digestFunction;
+        this.hashingFunction = hashingFunction;
+    }
+
+    /**
+     * Get accessions for a list of messages. It looks for the object's accessions in a repository, and it they don't
+     * exist, generate new ones, storing them in the repository
+     *
+     * @param messages
+     * @return
+     */
+    @Override
+    public Map<ACCESSION, MODEL> getAccessions(List<? extends MODEL> messages) {
+        Map<HASH, MODEL> hashToMessages = mapHashOfMessages(messages);
+        Map<HASH, ACCESSION> existingAccessions = dbService.getExistingAccessions(hashToMessages.keySet());
+        Map<HASH, MODEL> newMessages = filterNotExists(hashToMessages, existingAccessions);
+
+        Map<ACCESSION, MODEL> accessions = joinExistingAccessionsWithMessages(existingAccessions, hashToMessages);
+        if (!newMessages.isEmpty()) {
+            accessions.putAll(generateAccessions(newMessages));
+        }
+        return accessions;
+    }
+
+    /**
+     * Digest messages, hash them and map them. If Two messages have the same hash keep the first one.
+     *
+     * @param messages
+     * @return
+     */
+    private Map<HASH, MODEL> mapHashOfMessages(List<? extends MODEL> messages) {
+        return messages.stream().collect(Collectors.toMap(digestFunction.andThen(hashingFunction), e -> e, (r, o) -> r));
+    }
+
+    private Map<HASH, MODEL> filterNotExists(Map<HASH, MODEL> hashToMessages,
+                                             Map<HASH, ACCESSION> existingAccessions) {
+        return hashToMessages.entrySet().stream()
+                .filter(entry -> !existingAccessions.containsKey(entry.getKey()))
+                .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+    }
+
+    private Map<ACCESSION, MODEL> joinExistingAccessionsWithMessages(Map<HASH, ACCESSION> existingAccessions,
+                                                                     Map<HASH, MODEL> hashToMessages) {
+        return existingAccessions.values().stream().collect(Collectors.toMap(e -> e, e -> hashToMessages.get(e)));
+    }
+
+    private Map<ACCESSION, MODEL> generateAccessions(Map<HASH, MODEL> accessions) {
+        SaveResponse<ACCESSION, MODEL> response = dbService.save(accessionGenerator.generateAccessions(accessions));
+        accessionGenerator.postSave(response);
+        return response.getAllAccessionToMessage();
+    }
+
+    @Override
+    public Map<ACCESSION, ? extends MODEL> get(List<MODEL> accessionObjects) {
+        return dbService.findAllAccessionByMessageHash(getHashes(accessionObjects));
+    }
+
+    private List<HASH> getHashes(List<MODEL> accessionObjects) {
+        return accessionObjects.stream().map(digestFunction.andThen(hashingFunction)).collect(Collectors.toList());
+    }
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
```

Comment:
I don't think this class can made abstract, i think it is used. As I've said to Selva also, put the InitializingBean so that all the descending classes are suppossed to be injected and that the possibility to execute an initialization is something feasible even if you extend them. We can move the implements to the child class that really uses it. No strong oppinion in this, as it is a common practise in Spring.